pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,604 @@
+            // This is to ensure all the rewards are paid out according to the new inflation configuration from next block.
+            //
+            // If this was done in `on_initialize`, collator & treasury would receive incorrect rewards for that one block.
+            //
+            // This should be done as late as possible, to ensure all operations that modify issuance are done.
+            if Self::is_recalculation_in_next_block(now, &ActiveInflationConfig::<T>::get()) {
+                let (max_emission, config) = Self::recalculate_inflation(now);
+                ActiveInflationConfig::<T>::put(config.clone());
+
+                SafetyInflationTracker::<T>::mutate(|tracker| {
+                    tracker.cap.saturating_accrue(max_emission);
+                });
+
+                Self::deposit_event(Event::<T>::NewInflationConfiguration { config });
+            }
+        }
+
+        fn integrity_test() {
+            assert!(T::CycleConfiguration::periods_per_cycle() > 0);
+            assert!(T::CycleConfiguration::eras_per_voting_subperiod() > 0);
+            assert!(T::CycleConfiguration::eras_per_build_and_earn_subperiod() > 0);
+            assert!(T::CycleConfiguration::blocks_per_era() > 0);
+        }
+    }
+
+    impl<Moment, T: Config> OnTimestampSet<Moment> for Pallet<T> {
+        fn on_timestamp_set(_moment: Moment) {
+            let amount = Self::payout_block_rewards();
+
+            // Update the tracker, but no check whether an `issued` has exceeded `cap`.
+            // This can modified if needed, but these amounts are supposed to be small &
+            // collators need to be paid for producing the block.
+            // TODO: potential discussion topic for the review!
+            SafetyInflationTracker::<T>::mutate(|tracker| {
+                tracker.issued.saturating_accrue(amount);
+            });
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Used to force-set the inflation parameters.
+        /// The parameters must be valid, all parts summing up to one whole (100%), otherwise the call will fail.
+        ///
+        /// Must be called by `root` origin.
+        ///
+        /// Purpose of the call is testing & handling unforseen circumstances.
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::force_set_inflation_params())]
+        pub fn force_set_inflation_params(
+            origin: OriginFor<T>,
+            params: InflationParameters,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            ensure!(params.is_valid(), Error::<T>::InvalidInflationParameters);
+            InflationParams::<T>::put(params);
+
+            Self::deposit_event(Event::<T>::InflationParametersForceChanged);
+
+            Ok(().into())
+        }
+
+        /// Used to force-set the inflation configuration.
+        /// The parameters aren't checked for validity, since essentially anything can be valid.
```

Comment:
Since this would pass all validity checks and could go very wrong if not carefully configured. I suggest to:
1. Either remove it.
2. Or document it well and say: a. use force setting inflation params & force recalculation for most cases; b. this call is the last option and shouldn't be used without good tests on live fork network.

Reply:
There is no way to validate config because anything goes.
I will add the comments you suggested.

I'd just like to add as well that I want this calls just for testing & possible early issues.
I'm perfectly fine with removing this later - for now I'd like to keep it for testing.